File: client/dart/lib/src/record_id.dart
abstract class RecordId {
  @override
  String toString();

  factory RecordId.integer(int id) => _IntegerRecordId(id);
  factory RecordId.uuid(String id) => _UuidRecordId(id);

  static RecordId parse(String id) {
    final intId = int.tryParse(id);
    if (intId != null) {
      return RecordId.integer(intId);
    }
    return RecordId.uuid(id);
  }
}

class _IntegerRecordId implements RecordId {
  final int id;

  const _IntegerRecordId(this.id);

  @override
  String toString() => id.toString();

  @override
  bool operator ==(Object other) {
    if (other is _IntegerRecordId) return id == other.id;
    if (other is int) return id == other;
    return false;
  }

  @override
  int get hashCode => id.hashCode;
}

extension RecordIdExtInt on int {
  RecordId id() => _IntegerRecordId(this);
}

class _UuidRecordId implements RecordId {
  final String id;

  const _UuidRecordId(this.id);

  @override
  String toString() => id;

  @override
  bool operator ==(Object other) {
    if (other is _UuidRecordId) return id == other.id;
    if (other is String) return id == other;
    return false;
  }

  @override
  int get hashCode => id.hashCode;
}

extension RecordIdExtString on String {
  RecordId id() => _UuidRecordId(this);
}

File: client/dotnet/trailbase/RecordId.cs
namespace TrailBase;

/// <summary>Base for RecordId representations.</summary>
public abstract class RecordId {
  /// <summary>Serialize RecordId.</summary>
  public abstract override string ToString();

  /// <summary>Parses RecordIds.</summary>
  static public RecordId Parse(string id) {
    long value = 0;
    if (long.TryParse(id, out value)) {
      return new IntegerRecordId(value);
    }
    return new UuidRecordId(id);
  }
}

/// <summary>Integer record id.</summary>
public class IntegerRecordId : RecordId {
  long id { get; }

  /// <summary>Integer record id constructor.</summary>
  public IntegerRecordId(long id) {
    this.id = id;
  }

  /// <summary>Serialize RecordId.</summary>
  public override string ToString() => id.ToString();
}

/// <summary>UUID record id.</summary>
public class UuidRecordId : RecordId {
  Guid id { get; }

  /// <summary>UUID record id constructor.</summary>
  public UuidRecordId(Guid id) {
    this.id = id;
  }

  /// <summary>UUID record id constructor.</summary>
  public UuidRecordId(string id) {
    var bytes = System.Convert.FromBase64String(id.Replace('-', '+').Replace('_', '/'));
    this.id = new Guid(bytes);
  }

  /// <summary>Serialize UuidRecordId.</summary>
  public override string ToString() {
    var bytes = id.ToByteArray();
    return System.Convert.ToBase64String(bytes).Replace('+', '-').Replace('/', '_');
  }
}
